### DISCUSSION TITLE
Improved Google Maps KML to OSMAnd style GPX file converter including icon conversion

### DISCUSSION THREAD

**Discussion by tmusolf** [osmandapp/OsmAnd#17030] [Show and tell]
I did a big rewrite on a KML to OSMAnd GPX converter I wrote a few years back and thought it might be useful to the OSMAnd community.  It's written in Python.

Here's the link to the [Github repository](https://github.com/tmusolf/KMLtoOSMAndGPX)

[Update]
New version of this utility to import a GMap directly and convert to OSMAnd style GPX files.  Most GMap icons are also converted.  [GoogleMaptoOSMAndGPX](https://github.com/tmusolf/GoogleMapToOSMAndGPX)

I created this utility because I do all of my trip planning by creating a google my maps map.  I want to be able to use OSMAnd offline as many of the locations I travel have no cell coverage so it's important to be able to import these files into OSMAnd.  

The KMLtoOSMAndGPX utility converts waypoints and track elements from the KML file.  By default all tracks and waypoints are written to a single GPX file. There is an option to write one GPX file per KML layer/folder.  

Each waypoint will have it's icon translated according to a table in the program.  It will use a fixed color value or optionally the color from the KML file.  Description, lat, lon and elevation are all translated.

All tracks will have a fixed line width and transparency which can be specified with command line parameters.  Lat, lon and elevation of each track point is translated.  Descriptions are translated into the standard GPX description tag, but unfortunately OSMAnd does not display a description for individual tracks.  The OSMAnd feature for showing track arrows is set to off.  The show splits feature is disabled by default, but can be specified.

When importing the GPX files into OSMAnd you should import all tracks into a single file.  If you select the option for OSMAnd to create separate GPX files for each track, the initial track color and width doesn't function correctly.  

I admit I may have an incomplete understanding of how OSMAnd handles imported tracks and the app specific extensions tags it uses.   I'm unclear how OSMAnd determines it's initial track attributes in all cases.  It does appear that OSMAnd does not let you manage the appearance attributes of individual tracks contained within a single GPX file.  The show split interval feature does not reset the mileage with each track in a GPX file.  Instead the mileage shown on the interval markers is cumulative across all the tracks in the file.

Using the edit appearance menu will reset all the track color and line width values to the same setting, overriding what was specified in the GPX file.  Reset to original does not restore the original color or line width values.

Most of my maps have many tracks and waypoints.  For example, all the trails, trail junctions, POI's that I'm interested in for a national park.  Within OSMAnd I want to be able to turn visibility of these tracks and waypoints on/off as a single file, without having to manage many different GPX files.  Ideally, I could view and change the attributes of individual items within a single GPX file, but OSMAnd is still lacking in this area.

**Comment by ThomasMeister123** [comment]
Your links goes to "Didn’t find anything here!" (by clicking). Copy Linktext and paste works...

**Reply by tmusolf** [reply]
Thanks.  Fixed in original post.

**Comment by donnje** [comment]
if I understand correctly you can choose the color of the icons but not import all the customized ones in googlemap?

**Reply by tmusolf** [reply]
There are many google maps icons.  I didn't take the time to add all of them to the table contained in the code and map them to a corresponding OSMAnd icon.  It's easy to add any additional icons you need into the icon dictionary.

**Comment by mariush444** [comment]
Another converter
https://github.com/mariush444/gmapIcons2osmand

**Comment by tmusolf** [comment]
I've created a new utility, GoogleMapToOSMAndGPX, to converts directly from a google map into OSMAnd style GPX files.  It can be found here: [Git Repository](https://github.com/tmusolf/GoogleMapToOSMAndGPX/tree/main)

**Comment by mariush444** [comment]
@vshcherb 
There are some python scripts for osmand. Mainly converters.
But generally it is usefull for advanced users and not for everyone.
Do you think it is good idea to add posibility to run this scripts for users on osmand (map?) web page? Maybe it would be possibility to run scripts as a plugin? It would be great advantege of osmand for every user.
What is your opinion?

**Reply by tmusolf** [reply]
My planning workflow has always been create a google custom map as I research a trip.  It's much easier to work on a PC to do this.  Once I'm ready I'll move that data into OSMAnd.  I've always wished there was an OSMAnd app feature to smartly import google maps. Over the years I've written several versions of a KML to OSMAnd GPX converter to address this problem, all on Github.  Your [gmapicons2osmand](https://github.com/mariush444/gmapIcons2osmand) utility was based on [my first converter version](https://www.mail-archive.com/osmand@googlegroups.com/msg07817.html).

As you read comments in the OSMAnd forums you realize that people use the app in so many different ways.  I'm sure a google maps to OSMAnd converter would be more frequently used if it was a direct conversion without having to move files to/from specific locations.  As a plugin it could directly convert and then "import" a map into OSMAnd, no manual file moving required.  As a website tool it would be a bit like the useful [GPSVisualizer](https://www.gpsvisualizer.com/) site and create the GPX files that you'd then have to manually move into OSMAnd.  I have no idea how to approach either of these options.

I've posted several questions in the OSMAnd discussion forums trying to get a better understanding of OSMAnd handling of GPX files because there are features that still don't work the way I'd like, but I've never received answers.

BTW I just noticed the proposed icon conversion document you had posted.  When I get some time I'll incorporate that entire table into my converter's code.  Do you have any idea how to suggest new OSMAnd icons to the OSMAnd team to fill in some of the missing icons?

**Reply by mariush444** [reply]
> Do you have any idea how to suggest new OSMAnd icons to the OSMAnd team to fill in some of the missing icons?

No idea other than make an issue and ask.
Sometimes I informed them about "bugs" in icons and they correct them but when I ask for new general icons instead unique like Colloseum they said that they have no idea to which group icons should belong.

So maybe it is idea to ask for icon with example and say to which group should be added?

Other posibility is (maybe) to make fork a repository add icons and ask team to"reconnet" (I don't   remember propper name) to main tepository.

I did it like that with web docs of osmand and they accepted.

But in this issue I'm not good in graphic nor java :(

**Reply by tmusolf** [reply]
I think a good solution would be for OSMAnd to create an entirely new group of icons that map directly to gmap icons.  Leave the existing OSMAnd icons as they are.  There is no reason the two cannot exist together.

**Reply by mariush444** [reply]
> As a website tool it would be a bit like the useful [GPSVisualizer](https://www.gpsvisualizer.com/) site and create the GPX files that you'd then have to manually move into OSMAnd

After conversion gpx can be packed into osf. There is possible to create plugin that is not shown in menu osmand after importing a file. So there is no problem with planty of plugins in menu.

**Reply by tmusolf** [reply]
I've updated [GoogleMapToOSMAndGPX](https://github.com/tmusolf/GoogleMapToOSMAndGPX) utility icon translation table to convert almost all GMap icons to OSMAnd equivalents, according to your document.  There are also 2 windows executables, one is a GUI wrapper and the other is the utility itself.
